### PR TITLE
Generate parameters per each resource file

### LINKED ISSUE #356: Extract tool generating invalid parameters file to deploy
After extracting my APIM using the following parameters (**real values removed**):

```
{
  "sourceApimName": "removed",
  "destinationApimName": "removed",
  "resourceGroup": "removed",
  "fileFolder": "C:\\Source",
  "paramNamedValue": true,
  "paramApiLoggerId": true,
  "paramLogResourceId": true,
  "baseFileName": "removed"
}
```

I got this parameters.json file:

```
{
  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "ApimServiceName": {
      "value": "removed"
    },
    "NamedValues": {
      "value": {
        "removed-prop-name": "removed-value"
      }
    },
    "ApiLoggerId": {
      "value": {
        "my-api-name": {
          "applicationinsights": "resource-address"
        }
      }
    },
    "LoggerResourceId": {
      "value": {
        "azuremonitor": null,
        "PROPNAME": "resource-address"
      }
    }
  },
  "resources": []
}
```

Later when running my release pipeline on Azure DevOps I've got this error:

**Deployment template validation failed: 'The template parameters 'ApiLoggerId, LoggerResourceId' in the parameters file are not valid; they are not present in the original template and can therefore not be provided at deployment time. The only supported parameters for this template are 'ApimServiceName, NamedValues'. Please see https://aka.ms/arm-deploy/#parameter-file for usage details.'.**

Does anyone have any idea how to fix it? I've tried changing the schema to 2019-04-01, but got the same error.

### PR DESCRIPTION
**Pull request by f-alizada**
Generates parameter file per each resource template file. 
Closes #356

### CODE REVIEW

File: src/ArmTemplates/Commands/Configurations/ExtractorConsoleAppConfiguration.cs [hunk 1]
```diff
@@ -106,6 +106,9 @@ public class ExtractorConsoleAppConfiguration
         [Option(longName: "extractIdentityProviders", HelpText = "Extract identity providers from the service if applies")]
         public string ExtractIdentityProviders { get; set; }
 
+        [Option(longName: "parametersOutputDirectoryName ", HelpText = "Parameters output directory parameter, by default its parameters")]
```
Comment (DeagleGross) [inline src/ArmTemplates/Commands/Configurations/ExtractorConsoleAppConfiguration.cs:109]:
```suggestion
        [Option(longName: "parametersOutputDirectoryName ", HelpText = "Parameters output directory name, by default it is \"parameters\"")]
```

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 2]
```diff
@@ -459,6 +460,81 @@ public async Task<Template<GroupTemplateResources>> GenerateGroupsTemplateAsync(
...
+            Template<LoggerTemplateResources> loggersTemplate = null,
+            Template<BackendTemplateResources> backendsTemplate = null,
+            Template<AuthorizationServerTemplateResources> authorizationServersTemplate = null,
+            Template<NamedValuesResources> namedValuesTemplate = null,
+            Template<TagTemplateResources> tagTemplate = null,
+            Template<GroupTemplateResources> groupTemplate = null,
+            Template<IdentityProviderResources> identityProviderTemplate = null,
+            Template<SchemaTemplateResources> schemaTemplate = null,
+            Template<OpenIdConnectProviderResources> openIdConnectProviderTemplate = null,
+            Template<PolicyFragmentsResources> policyFragmentsTemplate = null
```
Comment (DeagleGross) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:481]:
```suggestion
            Template<PolicyFragmentsResources> policyFragmentsTemplate = null)
```

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 3]
```diff
@@ -459,6 +460,81 @@ public async Task<Template<GroupTemplateResources>> GenerateGroupsTemplateAsync(
...
+            )
+        {
+
```
Comment (DeagleGross) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:484]:
```suggestion
```

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 4]
```diff
@@ -459,6 +460,81 @@ public async Task<Template<GroupTemplateResources>> GenerateGroupsTemplateAsync(
             return templateParameters;
         }
 
+        public async Task GenerateResourceParametersFiles(
```
Comment (DeagleGross) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:463]: Do we really need a top level function here? What about generating and storing the parameters file for dedicated resource in it's special method?
Like
```
public async Task<Template<ApiTemplateResources>> GenerateApiTemplateAsync(
            string singleApiName,
            List<string> multipleApiNames,
            string baseFilesGenerationDirectory)
        {
            // normal extractor behaviour here
            
            var parametersFile = await this.parametersGenerator.GenerateParameters<ApiTemplateResources>(apiTemplates);
            
            if (parametersFile ...)
            {
                ....
            }
            
            this.logger.LogInformation("Finished generation of api template...");
            return apiTemplate;
        }
```

Comment (f-alizada) [reply src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:463]: I was thinking to go this way first, but here are some thoughts: 

1. In order to generate main parameters file we have to have resources already extracted (for example identityProviders, OpenIdConnectProviders)
2. The way introduced here: we are relying on only one method of generation of parameters (GenerateParametersTemplateAsync) from which all other parameters are generated. So, generation of parameters executes only ones, and the method just collects needed parameters without logic behind. 
3. In the suggested way we have to generate parameters for specific resource plus include the same logic into generation of main parameter template.

File: src/ArmTemplates/Commands/Configurations/ExtractorConsoleAppConfiguration.cs (hunk 1 above)
Comment (f-alizada) [reply src/ArmTemplates/Commands/Configurations/ExtractorConsoleAppConfiguration.cs]: Updated thank you!

**Review by DeagleGross — APPROVED** [review]

**Pull request merged**